app/uk/gov/hmrc/personaldetailsvalidation/views/pages/PersonalDetailsPage.scala

```diff
@@ -38,49 +38,50 @@ private[personaldetailsvalidation] class PersonalDetailsPage @Inject()(implicit
-      case _ => throw new IllegalStateException("Either of nino or postcode should be present")
-    }
+  private val formWithPostcode: Form[PersonalDetailsWithPostcode] = Form(mapping(
+    "firstName" -> mandatoryText("personal-details.firstname.required"),
+    "lastName" -> mandatoryText("personal-details.lastname.required"),
+    "postcode" -> postcodeValidation(),
+    "dateOfBirth" -> mandatoryLocalDate("personal-details")
+  )(PersonalDetailsWithPostcode.apply)(PersonalDetailsWithPostcode.unapply))
 
-  private val createPersonalDetailsData: PersonalDetails=> PersonalDetailsData ={
-    case PersonalDetailsWithNino(firstName, lastName, nino, dateOfBirth) => (firstName, lastName, Some(nino), dateOfBirth, None)
-    case PersonalDetailsWithPostcode(firstName, lastName, postcode, dateOfBirth) => (firstName, lastName, None, dateOfBirth, Some(postcode))
+  private def ninoValidation(): Mapping[Nino] = {
+    mandatoryText("personal-details.nino.required")
+      .verifying("personal-details.nino.invalid", nonEmptyString => Try(Nino(nonEmptyString.value)).isSuccess)
+      .transform[Nino](validatedNonEmptyNino => Nino(validatedNonEmptyNino.value), nino => NonEmptyString(nino.toString))
   }
 
-  private val ninoAndPostcodeMutuallyExclusive: PersonalDetailsData => Boolean = {
-    case (_, _, Some(_), _, Some(_)) => false
-    case (_, _, None, _, None) => false
-    case _ => true
+  private def postcodeValidation(): Mapping[NonEmptyString] = {
+    mandatoryText("personal-details.postcode.invalid").
+      verifying("personal-details.postcode.invalid", postcodeFormatValidation _)
   }
 
-  private val form: Form[PersonalDetails] = Form(mapping(
-    "firstName" -> mandatoryText("personal-details.firstname.required"),
-    "lastName" -> mandatoryText("personal-details.lastname.required"),
-    "nino" -> optionalText
-      .verifying("personal-details.nino.invalid", nonEmptyString => nonEmptyString.forall(nonEmptyString => Try(Nino(nonEmptyString.value)).isSuccess))
-      .transform[Option[Nino]](_.map(nonEmptyString => Nino(nonEmptyString.value)), _.map(nino => NonEmptyString(nino.value))),
-    "dateOfBirth" -> mandatoryLocalDate("personal-details"),
-    "postcode" -> optionalText
-  )(personalDetailsParser)(Some(_))
-    .verifying("personal-details.ninoOrPostcode.required", ninoAndPostcodeMutuallyExclusive)
-    .transform(createPersonalDetails, createPersonalDetailsData)
-  )
+  private def postcodeFormatValidation(postcode: NonEmptyString) =
+    postcode.value.matches("""([A-Za-z][A-HJ-Ya-hj-y]?[0-9][A-Za-z0-9]?|[A-Za-z][A-HJ-Ya-hj-y][A-Za-z])\s?[0-9][ABDEFGHJLNPQRSTUWXYZabdefghjlnpqrstuwxyz]{2}""")
 
-  def render(implicit completionUrl: CompletionUrl,
+  def render(showPostcodePage: Boolean = false)(implicit completionUrl: CompletionUrl,
              request: Request[_]): Html =
-    personal_details(form, completionUrl)
+    if(showPostcodePage) personal_details_postcode(formWithPostcode, completionUrl) else
```

Comment:
How about same template and conditionally rendering nino field or postcode field?

Reply:
How do we validate whether the current postcode or nino is required based on the condition selected? 

I did try creating two different partials for nino and postcode fields, but was having trouble creating a validation for both fields. e.g. saying nino is required if showing nino field, vice versa for postcode